proposed/resource-limit.md

```diff
@@ -0,0 +1,883 @@
+
+### Non-Goals
+
+#### Resource limit adoption in the BCL
+
+Two BCL types where adoption of the resource limiters are beneficial and relatively simple are:
+
+- System.Threading.Channels
+- System.IO.Pipelines
+
+Prototypes are illustrated in the [proof of concepts](#dotnetruntime-pocs) but are not commited to for .NET 6.
+
+#### HttpClient, Streams, Sockets
+
+Although we anticipate that there will be applications in many .NET areas and components such as HttpClient, Streams and Sockets, these scenarios will be explored but not committed for adoption in .NET 6 since they may be more complex. The adoption of the resource limiting APIs in these areas can happen independently of this proposal in the future.
+
+#### Partial Acquisition and Release
+
+Current design does not allow for partial acquisition or release of resources. In other words, either all of the request count is acquired or nothing is acquired. Likewise, when the operation is complete, all resources acquired via the limiter is released together. Though there are theoretically potential use cases for partial acquisitions and releases, no concrete examples has been identified in ASP.NET Core or .NET. In case these functionalities are needed, additional APIs can be added indepently from the APIs proposed here.
+
+For example, this will likely involve a modified `ResourceLease` (expand details for API)
+
+<details>
+
+```c#
+public struct ResourceLease : IDisposable
+{
+    // This represents whether resource lease acquisition was successful
+    public bool IsAcquired { get; }
+
+    // Number of acquired resources in the lease. Alternatively this can be stored on `State`
+    public long ResourceCount { get; }
+
+    // This represents additional metadata that can be returned as part of a call to Acquire/WaitAsync
+    // Potential uses could include a RetryAfter value or an error code.
+    public object? State { get; }
+
+    // `state` is the additional metadata, onDispose takes `state` as its argument.
+    public ResourceLease(long count, bool isAcquired, object? state, Action<ResourceLease>? onDispose);
+
+    // Release a specific amount of resources, if possible
+    public void Release(long releaseCount);
+
+    // Release remaining resources
+    public void Dispose();
+
+    // This static `ResourceLease` is used by rate limiters or in cases where `Acquire` returns false.
+    public static ResourceLease SuccessfulAcquisition = new ResourceLease(true, null, null);
+    public static ResourceLease FailedAcquisition = new ResourceLease(false, null, null);
```

Comment:
```suggestion
```

Reply:
I'll finish updating these in the morning.